### ISSUE TITLE
xlab & ylab getting cut off using ggplotly()

### ISSUE THREAD

**Issue by kidman007** [ropensci/plotly#510]
Hello,

I'm new to Plotly and am very excited to use it's flash visualizations and the tooltip. I'm running into an issue with my x & y labels getting cut off when I run the viz with ggplotly -- whereas it works fine when I just run the ggplot().

Below is the code and viz. I appreciate the help in advance!

Andrew

![image](https://cloud.githubusercontent.com/assets/12884043/13791063/5996f552-eab9-11e5-81d9-b6bdb48b97f3.png)

Here's my code:
```
testPlot <-
ggplot(TopCarePaths.Req,aes(x = calc_appears_offset, y = total.completed_sum/total.completed_and_expired_sum ) )+
  geom_point( aes( shape = is_form,  color = PrePostFlag, text = paste("Title:", title_full) ), alpha = 2/5, size = 5) + #dimensions of CarePath
  # geom_point( aes( text = paste("Title:", title_full) ) ) + # changes the tooltip
  # geom_rug(sides="b", position='jitter') + #adds rug, which counts number of points. Added so only applied to X-axis & spread out with jitter
  guides( alpha = FALSE ) + #hides unneccssary legend for hard coded transparancy
  facet_grid( organization_parent_name~. ) + #Facets graph horizontally
  scale_x_continuous( limits = c( -50, 50 ), breaks = seq( -50, 50, 5 ) ) + #controls X axis tick marks
  scale_y_continuous(labels = scales::percent) + #quickly changes y axis to %
  geom_vline( xintercept = 0 ) + #adds solid line to chart
  geom_text( aes( x = 0, label = "Surgery\n", y = .2), family = 'sans', angle=0,  color = 'gray34', size = 3, alpha = 1/5 ) + #puts text next to line for surg date
  theme_bw() + #changes to a simpler, easy to read theme
  theme( strip.background = element_blank() ) + #removes ugly gray from name area of facet
  ggtitle( 'Content Completion by Across Full CarePath' ) +
  ylab('Completion %') +
  xlab( 'Days Appears Offset from Surgery' )
ggplotly()
```

**Comment by bhive01** [comment]
I'm seeing this too on my plots lately, but it appears to be related to the window size/ratio. 
Here's a MRE:

```r
require(ggplot2)
require(plotly)

randomName <- function(n) {
	random.string <- rep(NA, n)
	randomizeString <- function(x) {
		a <-sample(letters, 1, replace = TRUE)
		return(a)
	}
	return(paste(c(sapply(random.string, randomizeString, simplify = TRUE)), collapse = ""))
}

xvar = replicate(10, randomName(15))

df1 <- data.frame(x = xvar, y = xvar, z = runif(10) )

g = ggplot(df1, aes(x,y,fill=z)) +
		geom_tile() 

ggplotly(g)
``` 

Squished, the axis name overlaps with the axis labels:
![screen shot 2016-03-15 at 2 20 16 pm](https://cloud.githubusercontent.com/assets/394973/13794468/268a69c8-eab9-11e5-9814-70f2c7253011.png)
nearly 1:1 everything looks ok
![screen shot 2016-03-15 at 2 20 24 pm](https://cloud.githubusercontent.com/assets/394973/13794470/2974217e-eab9-11e5-8f55-8b9024e316e7.png)
longer and the axis name goes off the end
![screen shot 2016-03-15 at 2 20 31 pm](https://cloud.githubusercontent.com/assets/394973/13794471/2c2f639c-eab9-11e5-9691-a5720d198fce.png)

**Comment by andy-ewing** [comment]
![screen shot 2016-03-22 at 1 32 10 pm](https://cloud.githubusercontent.com/assets/14897539/13967056/8e1147aa-f032-11e5-8f87-67b9a3533258.png)

I'm having similar issues since the latest updates to R, RStudio, and plotly packages. It may have something to do with the scales package?

**Comment by jasdumas** [comment]
I'm experiencing the same issue with making an unstacked bar chart and the y-axis label is getting cut-off similarly to @kidman007. I have tried some `theme()` adjustments within `ggplot` but with no success. 

Has there been any solutions yet?

**Comment by timelyportfolio** [comment]
Looking at it now, but I am still not as familiar with the code as I would like to be.  It appears these [lines](https://github.com/ropensci/plotly/blob/master/R/ggplotly.R#L403-L426) are most likely to hold the answer.  I'll dig deeper.

**Comment by timelyportfolio** [comment]
When I reproduce the plot with the example from @bhive01, I get an annotation for `x` with [`yref`](https://plot.ly/javascript/reference/#layout-annotations-yref) equal to `paper` which means the annotation will be positioned with reference to the chart `paper` box on a scale `(0,1)`.  The negative value provided (`-0.07990868`) will scale linearly and inversely based on the size of the plot, which explains why when the plot is small, the `x` is too high, and when the plot is large, `x` is too low/off the screen.

I'll ping @cpsievert to see if there is an ideal solution, or if we should resort to a hack.

**Comment by cpsievert** [comment]
Yea, I'm not sure the best way to approach this problem. What I really need is a way to place annotations on a (0-1) scale of the entire plot, or on the absolute pixel scale. Any idea if that would be easy to add to plotly.js?

**Comment by timelyportfolio** [comment]
@cpsievert, do you remember why in [lines](https://github.com/ropensci/plotly/blob/master/R/ggplotly.R#L402-L403) this comment was made?

```
        # draw axis titles as annotations
        # (plotly.js axis titles aren't smart enough to dodge ticks & text)
```

I am guessing facets, but could be way off here.

Let's hack on the example given above by defining `$x$layout$xaxis$title`.

```
require(ggplot2)
require(plotly)

randomName <- function(n) {
  random.string <- rep(NA, n)
  randomizeString <- function(x) {
    a <-sample(letters, 1, replace = TRUE)
    return(a)
  }
  return(paste(c(sapply(random.string, randomizeString, simplify = TRUE)), collapse = ""))
}

xvar = replicate(10, randomName(15))

df1 <- data.frame(x = xvar, y = xvar, z = runif(10) )

g = ggplot(df1, aes(x,y,fill=z)) +
  geom_tile() 

gpw <- as.widget(ggplotly(g))

gpw$x$layout$xaxis$title = "plotly's x"

gpw
```

![image](https://cloud.githubusercontent.com/assets/837910/14749209/cd079ac2-0883-11e6-9d06-286506280da4.png)

**Comment by jasdumas** [comment]
I've tried hacking with gpw$sizingPolicy$padding = 10 but it's still cut-off. For reference my plotly viz is on shiny where the yaxis is being cut-off by the navbarPage. 

However I just downloaded the plot to .png from the button and the yaxis is **completely normal** and not cut-off :confounded:

**Comment by cpsievert** [comment]
@timelyportfolio that's right, I'm drawing titles as annotations because x/y titles are always relative the plot in ggplot2, not each axis.

(in that comment, I think I meant to say "_annotations_ aren't smart enough to dodge ticks & text"

**Comment by joshmyzie2** [comment]
Probably obvious, but an easy/crude workaround is to append "\n " to your x label and prepend "\n " to your y label.  That should push your labels into the visible region.

**Comment by timelyportfolio** [comment]
Yes, but will worsen the situation when the plot is small.  @cpsievert, do you have an example where the real plotly axis labels don't work?

**Comment by cpsievert** [comment]
```r
qplot(data = mtcars, x = mpg, y = wt, facets = ~vs)
```

**Comment by timelyportfolio** [comment]
Also, mentioned in this new StackOverflow http://stackoverflow.com/questions/36988379/adjust-axis-positions-ggplot2-facets.

**Comment by drcrook1** [comment]
I'm having the same issues: http://stackoverflow.com/questions/36988379/adjust-axis-positions-ggplot2-facets.

**Comment by timelyportfolio** [comment]
@cpsievert, if no facets, should we change the behavior to use `plotly` defaults?

**Comment by cpsievert** [comment]
@timelyportfolio yea, that's a good point, a pull request would be most welcome :)

**Comment by timelyportfolio** [comment]
@cpsievert I took a first stab at a pull request to make this change.  I'm not sure that I got the axis title formatting correct.  I tried to still use your margin calculations for both facetted and non-facetted plots.  I know you will be much more familiar with the edge cases.  I would love your help thinking through the best way to test.

**Comment by cpsievert** [comment]
This should now be fixed for non-faceted plots, but we'll need a more general solution for facets.

**Comment by analca3** [comment]
Hi! I have this issue with a single plot, not faced. I don't have legend and the ylab is partial missed.

![image](https://cloud.githubusercontent.com/assets/3939991/15342349/f115ed72-1c94-11e6-8aec-72b55505561e.png)

This is the original plot, without `ggplotly(gg)` call.

![image](https://cloud.githubusercontent.com/assets/3939991/15342405/3309566a-1c95-11e6-97f3-5f8ac6f05d6b.png)

And this is the code

```
gg <- ggplot(data_replay, aes(Frame)) +
geom_line(aes(y=Minerals, colour = 'Ganador' )) +
geom_line(aes(y=EnemyMinerals, colour = 'Perdedor' )) +
ggtitle("Minerales") +
theme(legend.title=element_blank())

gg

ggplotly(gg)
```

My plotly version is 

```
> packageVersion("plotly")
[1] ‘3.4.13’
```

What can I do? Thank you!

**Comment by NattyE** [comment]
We seem to be seeing some of the plot being cut off (in our case the top of a bar chart) - possibly as a result of this - when used in Shiny. I can try to make a reproducible example if needed, but I don't have the source code.

**Comment by sulaxd** [comment]
try setting ` layout(margin = list(b = 90))`

**Comment by cpsievert** [comment]
Fixed via 9afbb5da88dd510c3aa44c855d8fb72c4e4b718e

**Comment by ldecicco-USGS** [comment]
I was having these margin issues, so I tried installing via github to get this fix. 

Now, I get an error:
```
library(ggplot2)
library(plotly)
df <- data.frame(x=1:10,y=1:10)
x2 <- ggplot(data = df) +
geom_point(aes(x=x,y=y))
ggplotly(x2)
Error in gg2list(p, width = width, height = height, tooltip = tooltip,  : 
  attempt to apply non-function
```

So, no problem, I'll just in-install this dev version of plotly and re-install via CRAN....but after tha, I was *still* getting that error.

I had a dev version of ggplot2 installed as well apparently (I was testing my code for the new updates that are coming out soon in ggplot2). Once I installed ggplot2 from CRAN, then that error finally went away.

I don't know if it's related to:
https://github.com/ropensci/plotly/issues/233
but might be.

So, I still have the margin issues...but admittedly didn't install the fix...but there might be some other issue with the fix....just FYI
